tests/telemetry/test_telemetry.py

```diff
@@ -240,3 +246,119 @@ def test_virtualdb_table_streaming(duthosts, rand_one_dut_hostname, ptfhost, loc
     assert_equal(len(re.findall('Max update count reached 3', result)), 1, "Streaming update count in:\n{0}".format(result))
     assert_equal(len(re.findall('name: "Ethernet0"\n', result)), 4, "Streaming updates for Ethernet0 in:\n{0}".format(result)) # 1 for request, 3 for response
     assert_equal(len(re.findall('timestamp: \d+', result)), 3, "Timestamp markers for each update message in:\n{0}".format(result))
+
+
+def run_gnmi_client(ptfhost, dut_ip):
+    """Runs gNMI client in the corresponding PTF docker.
+
+    Args:
+        pfthost: PTF docker binding to the selected DuT.
+        dut_ip: Mgmt IP of DuT.
+
+    Returns:
+        None.
+    """
+    gnmi_cli_cmd = 'python /gnxi/gnmi_cli_py/py_gnmicli.py -g -t {0} -p {1} -m subscribe  --trigger_mem_spike\
+                    -x DOCKER_STATS -xt STATE_DB -o "ndastreamingservertest"'.format(dut_ip, TELEMETRY_PORT)
+    logger.info("Starting gNMI client command in PTF docker: {}".format(gnmi_cli_cmd))
+    cmd_result = ptfhost.shell(gnmi_cli_cmd)
+    exit_code = cmd_result["rc"]
+    pytest_assert(exit_code == 0, "Failed to run gNMI client in PTF docker!")
+
+
+def terminate_gnmi_client(ptfhost):
+    """Terminate the gNMI client running in the PTF docker.
+
+    Args:
+        pfthost: PTF docker binding to the selected DuT.
+
+    Returns:
+        None.
+    """
+    gnmi_client_pid = ""
+
+    ps_cmd = "ps -aux | grep 'python /gnxi/gnmi_cli_py/py_gnmicli' | grep -v grep"
```

Comment:
>grep [](http://example.com/codeflow?start=24&length=4)

It is trick to grep and kill. Why not keep it running in foreground in one ptfhost connection, and shutdown the connection or task/thread later.

Reply:
I updated the python gNMI client (https://github.com/lguohan/gnxi/blob/64b7b03c42f3eb8ed5cffa7b58f1c0b58e6d8c40/gnmi_cli_py/py_gnmicli.py#L521) and it will be closed automatically if telemetry server was unavailable such as telemetry container was restarted by Monit.